man/man8/lpd.8
.\" Copyright (c) 1983 Regents of the University of California.
.\" All rights reserved.  The Berkeley software License Agreement
.\" specifies the terms and conditions for redistribution.
.\"
.\" @(#)lpd.8 1.1 92/07/30 SMI; from UCB 4.3 BSD 6.3
.\"
.TH LPD 8 "5 December 1989"
.SH NAME
lpd \- printer daemon
.SH SYNOPSIS
.B /usr/lib/lpd
[
.B \-l
] [
.B \-L
.I logfile
] [
.I port#
]
.SH DESCRIPTION
.IX  "lpd command"  ""  "\fLlpd\fP \(em line printer daemon"
.IX  printer "daemon \(em \fLlpd\fP"
.IX  "line printer daemon"  ""  "line printer daemon \(em \fLlpd\fP"
.LP
.B lpd
is the line printer daemon (spool area handler).
It is usually invoked at boot time from the
.BR rc (8)
script,  making a single pass through the
.BR printcap (5)
file to find out about the existing printers and printing any files
left after a crash.
It then accepts requests to print files in a queue,
transfer files to a spooling area, display a queue's status, or remove
jobs from a queue.
In each case, it forks a child process
for each request, and continues to listen for subsequent requests.
.LP
The Internet port number used to communicate
with other processes is usually obtained with
.BR getservent (3N),
but can be specified with the
.IR port#
argument.
.LP
If a file cannot be opened, an error message is logged using the
.SB LOG_LPR
facility of
.BR syslog (3).
.B lpd
will try up to 20 times
to reopen a file it expects to be there, after which it proceeds
to the next file or job.
.SH OPTIONS
.TP
.B \-l
Log valid requests received from the network.
This can be useful for debugging purposes.
.TP
.BI \-L " logfile"
Change the file used for writing error conditions to
.I logfile.
The default is to report a message using the
.BR syslog (3)
facility.
.SH OPERATION
.SS Access Control
.LP
Access control is provided by two means.
First, all requests must come from
one of the machines listed in either the file
.B /etc/hosts.equiv
or
.BR /etc/hosts.lpd .
(This latter file is in
.BR hosts.equiv (5)
format.)
Second, if the
.B rs
capability is specified in the
.B printcap
entry,
.BR lpr (1)
requests are only be honored for users with accounts on the
printer host.
.SS Lock File
The
.B lock
file in each spool directory is used to prevent multiple daemons from
becoming active, and to store information about the daemon process for
.BR lpr (1),
.BR lpq (1),
and
.BR lprm (1).
.LP
.B lpd
uses
.BR flock (2)
to provide exclusive access to the lock file and to prevent multiple
daemons from becoming active simultaneously.
If the daemon should be killed
or die unexpectedly, the lock file need not be removed.
The lock file is kept in a readable
.SM ASCII
form and contains two lines.
The first is the process id of the daemon
and the second is the control
file name of the current job being printed.
The second line is updated to
reflect the current status of
.B lpd
for the programs
.BR lpq (1)
and
.BR lprm (1).
.SS Control Files
After the daemon has successfully set
the lock, it scans the directory
for files beginning with
.BR cf .
Lines in each
.I cf
file specify files to be printed or non-printing actions to be
performed.  Each such line begins with a key character
that indicates what to do with the remainder of the line.
.RS
.PD 0
.TP
.B J
Job name to print on the burst page.
.TP
.B C
Classification line on the burst page.
.TP
.B L
Literal.  This line contains identification information from
the password file, and causes a burst page to be printed.
.TP
.B T
Title string for page headings printed by
.BR pr (1V).
.TP
.B H
Hostname of the machine where
.BR lpr (1)
was invoked.
.TP
.B P
Person.  Login name of the person who invoked
.BR lpr (1).
This is used to verify ownership by
.BR lprm (1).
.TP
.B M
Send mail to the specified user when the current print job completes.
.TP
.B f
Formatted File, the name of a file to print that is already formatted.
.TP
.B l
Like
.BR f ,
but passes control characters along, and does not make page breaks.
.TP
.B p
Name of a file to print using
.BR pr (1V)
as a filter.
.TP
.B t
Troff File.  The file contains
.BR troff (1)
output (cat phototypesetter commands).
.TP
.B n
Ditroff File.  The file contains device independent troff output.
.TP
.B d
.SM DVI
File.  The file contains
.SM T\v'.2v'E\v'-.2v'X
output (\s-1DVI\s0 format from Stanford).
.TP
.B g
Graph File.  The file contains data produced by
.BR plot (3X).
.TP
.B c
Cifplot File. The file contains data produced by
.IR cifplot .
.TP
.B v
The file contains a raster image.
.TP
.B r
The file contains text data with
.SM FORTRAN
carriage control characters.
.TP
.B 1
Troff Font R. The name of a font
file to use instead of the default.
.TP
.B 2
Troff Font I. The name of the font file
to use instead of the default.
.TP
.B 3
Troff Font B. The name of the font
file to use instead of the default.
.TP
.B 4
Troff Font S. The name of the font file
to use instead of the default.
.TP
.B W
Width. Changes the page width (in characters) used by
.BR pr (1V)
and the text filters.
.TP
.B I
Indent.  Specify the number of characters
by which to indent the output.
.TP
.B U
Unlink.  The name of file to remove upon completion of printing.
.TP
.B N
Filename.  The name of the file being printed, or a blank
for the standard input (when
.BR lpr (1)
is invoked in a pipeline).
.PD
.RE
.SS Data Files
.LP
When a file is spooled for printing,
the contents are copied into a data file
in the spool directory.
Data file names begin with
.BR df .
When
.B lpr
is called with the
.B -s
option, the control files contain a symbolic link
to the actual file, and no data files are created.
.SS Minfree File
The file
.I minfree
in each spool directory contains the
number of kilobytes to leave free
so that the line printer queue won't completely fill the disk.
.SH FILES
.PD 0
.TP 20
.B /etc/printcap
printer description file
.TP
.B /var/spool/*
spool directories
.TP
.B /var/spool/*/minfree
minimum free space to leave
.TP
.B /dev/lp*
line printer devices
.TP
.B /dev/printer
socket for local requests
.TP
.B /etc/hosts.equiv
hosts allowed equivalent host access
.TP
.B /etc/hosts.lpd
hosts allowed printer access only
.PD
.SH "SEE ALSO"
.BR lpq (1),
.BR lpr (1),
.BR lprm (1),
.BR hosts (5),
.BR hosts.equiv (5),
.BR printcap (5),
.BR lpc (8),
.BR pac (8)


ucb/lpr/lpd.c
/*
 * Copyright (c) 1983 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1983 Regents of the University of California.\n\
 All rights reserved.\n";
#endif not lint

#ifndef lint
static char sccsid[] = "@(#)lpd.c 1.1 92/07/30 SMI"; /* from UCB 5.4 5/6/86 */
#endif not lint

/*
 * lpd -- line printer daemon.
 *
 * Listen for a connection and perform the requested operation.
 * Operations are:
 *	\1printer\n
 *		check the queue for jobs and print any found.
 *	\2printer\n
 *		receive a job from another machine and queue it.
 *	\3printer [users ...] [jobs ...]\n
 *		return the current state of the queue (short form).
 *	\4printer [users ...] [jobs ...]\n
 *		return the current state of the queue (long form).
 *	\5printer person [users ...] [jobs ...]\n
 *		remove jobs from the queue.
 *
 * Strategy to maintain protected spooling area:
 *	1. Spooling area is writable only by daemon and spooling group
 *	2. lpr runs setuid root and setgrp spooling group; it uses
 *	   root to access any file it wants (verifying things before
 *	   with an access call) and group id to know how it should
 *	   set up ownership of files in the spooling area.
 *	3. Files in spooling area are owned by root, group spooling
 *	   group, with mode 660.
 *	4. lpd, lpq and lprm run setuid daemon and setgrp spooling group to
 *	   access files and printer.  Users can't get to anything
 *	   w/o help of lpq and lprm programs.
 */

#include "lp.h"

#include <sys/ioctl.h>

int	lflag;				/* log requests flag */

void	reapchild();
void	mcleanup();
void	exit();

main(argc, argv)
	int argc;
	char **argv;
{
	int f, funix, finet, options, fromlen;
	fd_set defreadfds;
	struct sockaddr_un s_un, fromunix;
	struct sockaddr_in sin, frominet;
	int omask, lfd;

	gethostname(host, sizeof host);
	name = argv[0];

	options = 0;
	while (--argc > 0) {
		argv++;
		if (argv[0][0] == '-')
			switch (argv[0][1]) {
			case 'd':
				options |= SO_DEBUG;
				break;
			case 'l':
				lflag++;
				break;
			}
	}

#ifndef DEBUG
	/*
	 * Set up standard environment by detaching from the parent.
	 */
	if (fork())
		exit(0);
	for (f = 0; f < 5; f++)
		(void) close(f);
	(void) open("/dev/null", O_RDONLY);
	(void) open("/dev/null", O_WRONLY);
	(void) dup(1);
	f = open("/dev/tty", O_RDWR);
	if (f > 0) {
		ioctl(f, TIOCNOTTY, 0);
		(void) close(f);
	}
#endif

	openlog("lpd", LOG_PID, LOG_LPR);
	(void) umask(0);
	lfd = open(MASTERLOCK, O_WRONLY|O_CREAT, 0644);
	if (lfd < 0) {
		syslog(LOG_ERR, "%s: %m", MASTERLOCK);
		exit(1);
	}
	if (flock(lfd, LOCK_EX|LOCK_NB) < 0) {
		if (errno == EWOULDBLOCK)	/* active deamon present */
			exit(0);
		syslog(LOG_ERR, "%s: %m", MASTERLOCK);
		exit(1);
	}
	ftruncate(lfd, (off_t)0);
	/*
	 * write process id for others to know
	 */
	sprintf(line, "%u\n", getpid());
	f = strlen(line);
	if (write(lfd, line, f) != f) {
		syslog(LOG_ERR, "%s: %m", MASTERLOCK);
		exit(1);
	}
	signal(SIGCHLD, reapchild);
	/*
	 * Restart all the printers.
	 */
	startup();
	(void) unlink(SOCKETNAME);
	funix = socket(AF_UNIX, SOCK_STREAM, 0);
	if (funix < 0) {
		syslog(LOG_ERR, "socket: %m");
		exit(1);
	}
#define	mask(s)	(1 << ((s) - 1))
	omask = sigblock(mask(SIGHUP)|mask(SIGINT)|mask(SIGQUIT)|mask(SIGTERM));
	signal(SIGHUP, mcleanup);
	signal(SIGINT, mcleanup);
	signal(SIGQUIT, mcleanup);
	signal(SIGTERM, mcleanup);
	s_un.sun_family = AF_UNIX;
	strcpy(s_un.sun_path, SOCKETNAME);
	if (bind(funix, (struct sockaddr *)&s_un,
			strlen(s_un.sun_path) + 2) < 0) {
		syslog(LOG_ERR, "ubind: %m");
		exit(1);
	}
	sigsetmask(omask);
	FD_ZERO(&defreadfds);
	FD_SET(funix, &defreadfds);
	listen(funix, 5);
	finet = socket(AF_INET, SOCK_STREAM, 0);
	if (finet >= 0) {
		struct servent *sp;

		if (options & SO_DEBUG)
			if (setsockopt(finet, SOL_SOCKET, SO_DEBUG,
					(char*)NULL, 0) < 0) {
				syslog(LOG_ERR, "setsockopt (SO_DEBUG): %m");
				mcleanup();
			}
		sp = getservbyname("printer", "tcp");
		if (sp == NULL) {
			syslog(LOG_ERR, "printer/tcp: unknown service");
			mcleanup();
		}
		bzero((char *)&sin, sizeof (struct sockaddr_in));
		sin.sin_family = AF_INET;
		sin.sin_port = sp->s_port;
		if (bind(finet, (struct sockaddr *)&sin, sizeof sin) < 0) {
			syslog(LOG_ERR, "bind: %m");
			mcleanup();
		}
		FD_SET(finet, &defreadfds);
		listen(finet, 5);
	}
	/*
	 * Main loop: accept, do a request, continue.
	 */
	for (;;) {
		int domain, nfds, s;
		fd_set readfds;

		readfds = defreadfds;
		nfds = select(20, &readfds, (fd_set *)0, (fd_set *)0,
				(struct timeval*)0);
		if (nfds <= 0) {
			if (nfds < 0 && errno != EINTR)
				syslog(LOG_WARNING, "select: %m");
			continue;
		}
		if (FD_ISSET(funix, &readfds)) {
			domain = AF_UNIX, fromlen = (sizeof fromunix);
			s = accept(funix, (struct sockaddr *)&fromunix,
					&fromlen);
		} else if (FD_ISSET(finet, &readfds)) {
			domain = AF_INET, fromlen = (sizeof frominet);
			s = accept(finet, (struct sockaddr *)&frominet,
					&fromlen);
		}
		if (s < 0) {
			if (errno != EINTR)
				syslog(LOG_WARNING, "accept: %m");
			continue;
		}
		if (fork() == 0) {
			signal(SIGCHLD, SIG_IGN);
			signal(SIGHUP, SIG_IGN);
			signal(SIGINT, SIG_IGN);
			signal(SIGQUIT, SIG_IGN);
			signal(SIGTERM, SIG_IGN);
			(void) close(funix);
			(void) close(finet);
			dup2(s, 1);
			(void) close(s);
			if (domain == AF_INET)
				chkhost(&frominet);
			else
				from_remote = 0;
			doit();
			exit(0);
		}
		(void) close(s);
	}
}

static void
reapchild()
{
	union wait status;

	while (wait3(&status, WNOHANG, (struct rusage *)0) > 0)
		;
}

static void
mcleanup()
{
	if (lflag)
		syslog(LOG_INFO, "exiting");
	unlink(SOCKETNAME);
	exit(0);
}

/*
 * Stuff for handling job specifications
 */
char	*user[MAXUSERS];	/* users to process */
int	users;			/* # of users in user array */
int	requ[MAXREQUESTS];	/* job number of spool entries */
int	requests;		/* # of spool requests */
char	*person;		/* name of person doing lprm */

static char	fromb[MAXHOSTNAMELEN + 1]; /* client's machine name */
static char	cbuf[BUFSIZ];	/* command line buffer */
static char	*cmdnames[] = {
	"null",
	"printjob",
	"recvjob",
	"displayq short",
	"displayq long",
	"rmjob"
};

static
doit()
{
	register char *cp;
	register int n;

	for (;;) {
		cp = cbuf;
		do {
			if (cp >= &cbuf[(sizeof cbuf) - 1])
				fatal("Command line too long");
			if ((n = read(1, cp, 1)) != 1) {
				if (n < 0)
					fatal("Lost connection");
				return;
			}
		} while (*cp++ != '\n');
		*--cp = '\0';
		cp = cbuf;
		if (lflag) {
			if (*cp >= '\1' && *cp <= '\5')
				syslog(LOG_INFO, "%s requests %s %s",
					from, cmdnames[*cp], cp+1);
			else
				syslog(LOG_INFO, "bad request (%d) from %s",
					*cp, from);
		}
		switch (*cp++) {
		case '\1':	/* check the queue and print any jobs there */
			printer = cp;
			printjob();
			break;
		case '\2':	/* receive files to be queued */
			printer = cp;
			recvjob();
			break;
		case '\3':	/* display the queue (short form) */
		case '\4':	/* display the queue (long form) */
			printer = cp;
			while (*cp) {
				if (*cp != ' ') {
					cp++;
					continue;
				}
				*cp++ = '\0';
				while (isspace(*cp))
					cp++;
				if (*cp == '\0')
					break;
				if (isdigit(*cp)) {
					if (requests >= MAXREQUESTS)
						fatal("Too many requests");
					requ[requests++] = atoi(cp);
				} else {
					if (users >= MAXUSERS)
						fatal("Too many users");
					user[users++] = cp;
				}
			}
			displayq(cbuf[0] - '\3');
			exit(0);
		case '\5':	/* remove a job from the queue */
			printer = cp;
			while (*cp && *cp != ' ')
				cp++;
			if (!*cp)
				break;
			*cp++ = '\0';
			person = cp;
			while (*cp) {
				if (*cp != ' ') {
					cp++;
					continue;
				}
				*cp++ = '\0';
				while (isspace(*cp))
					cp++;
				if (*cp == '\0')
					break;
				if (isdigit(*cp)) {
					if (requests >= MAXREQUESTS)
						fatal("Too many requests");
					requ[requests++] = atoi(cp);
				} else {
					if (users >= MAXUSERS)
						fatal("Too many users");
					user[users++] = cp;
				}
			}
			rmjob();
			break;
		}
		fatal("Illegal service request");
	}
}

/*
 * Make a pass through the printcap database and start printing any
 * files left from the last time the machine went down.
 */
static
startup()
{
	char buf[BUFSIZ];
	register char *cp;
	int pid;

	printer = buf;
	/*
	 * Restart the daemons.
	 */
	while (getprent(buf) > 0) {
		for (cp = buf; *cp; cp++)
			if (*cp == '|' || *cp == ':') {
				*cp = '\0';
				break;
			}
		if ((pid = fork()) < 0) {
			syslog(LOG_WARNING, "startup: cannot fork");
			mcleanup();
		}
		if (!pid) {
			endprent();
			printjob();
		}
	}
	printer = NULL;
}

#define DUMMY ":nobody::"

/*
 * Check to see if the from host has access to the line printer.
 */
static
chkhost(f)
	struct sockaddr_in *f;
{
	register struct hostent *hp;
	register FILE *hostf;
	register char *cp, *sp;
	char ahost[MAXHOSTNAMELEN + 1];
	int first = 1;
	extern char *inet_ntoa();
	char domain[MAXHOSTNAMELEN + 1];
	int baselen = -1;

	from = NULL;
	f->sin_port = ntohs(f->sin_port);
	if (f->sin_family != AF_INET || f->sin_port >= IPPORT_RESERVED)
		fatal("Malformed from address");
	hp = gethostbyaddr((char *)&f->sin_addr, sizeof (struct in_addr),
				f->sin_family);
	if (hp == 0)
		fatal("Host name for your address (%s) unknown",
			inet_ntoa(f->sin_addr));

	strcpy(fromb, hp->h_name);
	from = fromb;
	from_remote = (strcmp(from, host) != 0);
	if (!from_remote)
		return;

	getdomainname(domain, sizeof domain);
	sp = fromb;
	cp = ahost;
	while (*sp) {
		if (*sp == '.') {
			if (baselen == -1)
				baselen = sp - fromb;
			*cp++ = *sp++;
		} else {
			*cp++ = isupper(*sp) ? tolower(*sp++) : *sp++;
		}
	}
	*cp = '\0';
	hostf = fopen("/etc/hosts.equiv", "r");
again:
	if (hostf) {
		if (!_validuser(hostf, ahost, DUMMY, DUMMY, baselen)) {
			(void) fclose(hostf);
			return;
		}
		(void) fclose(hostf);
	}
	if (first == 1) {
		first = 0;
		hostf = fopen("/etc/hosts.lpd", "r");
		goto again;
	}
	fatal("Your host does not have line printer access");
}
